### DISCUSSION TITLE
Custom Modulus & External R1CS

### DISCUSSION THREAD

**Discussion by BeratOz01** [Consensys/gnark#595] [Q&A] [answered]
Hello, firstly, thank you for this beautiful library!

I have two questions:

Firstly, I couldn't figure out how to use mod operation with gnark. In circom:

Var xer = balance % r_prime; 
What is the equivalent of this in the gnark circuit?

Secondly, I have written and compiled my circuit in circom, now I have my r1cs and witness. Is it possible to use gnark only as a prover and also can I generate my witness from my r1cs (or sym etc.) of circom circuit, using gnark?

Thank you!!

**Comment by ivokub** [comment, marked as answer]
See https://pkg.go.dev/github.com/consensys/gnark@v0.8.0/std/math/emulated.

Yes, it is possible. We don't have transpiler from circom, but you can build R1CS circuit manually. See for example https://www.notamonadtutorial.com/how-to-use-the-consenyss-gnark-zero-knowledge-proof-library-and-disclosure-of-a-ddos-bug/

**Reply by BeratOz01** [reply]
Thank you for fast reply.
Actually, we already checked this `math/emulated` package but still couldn't figured out how to take modulus of 2 frontend.Variable. (circuit.X (mod circuit.Y)). Do you have any example for this?

**Reply by furkan-ux** [reply]
Yeah, all of the functions inside the `/emulated` takes *big.int and others. This is where I also stuck

**Reply by ivokub** [reply]
We don't have only a mod operation. I'm lacking context, what are you trying to achieve? I'm quite certain that you are better off with using field emulation gadget.

**Reply by furkan-ux** [reply]
![Ekran Resmi 2023-03-22 19 14 53](https://user-images.githubusercontent.com/69128891/226969093-63167b36-486f-4c96-864b-b69f3cd70e6f.png)
Just a simple mod?

**Reply by ivokub** [reply]
> We don't have only a mod operation. I'm lacking context, what are you trying to achieve? I'm quite certain that you are better off with using field emulation gadget.

For example, can you give a pseudocode of the computation?

**Reply by furkan-ux** [reply]
> > We don't have only a mod operation. I'm lacking context, what are you trying to achieve? I'm quite certain that you are better off with using field emulation gadget.
> 
> For example, can you give a pseudocode of the computation?

Just sent you a screenshot above. I need to perform a few mod operations. Computations must be private.

**Reply by furkan-ux** [reply]
@ivokub If I use math/emulated`  Do my inputs stay private when doing modulus? (as far as I understand, all other types except frontend.Variable are public)

And what I am trying to achieve is basically z := api.Mod(circuit.X, circuit.Y) // X = 10, Y = 3, then Z = 1

**Reply by BeratOz01** [reply]
@ivokub Any updates?

**Reply by ivokub** [reply]
We do not have it. You have to understand that in circuit we work in a finite field (so all variables are modulus the scalar field of the curve). So the result after the mod operation is still in the native field. And what could happen if you multiply two small values (lets say (r-1)/2, where r is the modulus of the scalar field) you overflow the native field. Then when you run api.Mod you get unexpected result (because the temp variable is already overflow the scalar field and reduced there).

That is why you would in general want to use `std/math/emulated`. Using that gadget we use a multi-precision arithmetic to ensure there is no overflow in the underlying native field. And I think the examples there are sufficient to start.

**Reply by ivokub** [reply]
And yes, using `std/math/emulated` the variables stay private if you initialise the elements from private vars.

**Reply by furkan-ux** [reply]
Thank you!!

**Reply by ivokub** [reply]
Another approach, if you want to do only modulus for the small example is to use hint. But then you have to be careful that the hint output is constrained. In pseudocode:
1) instead of `el % modulus = res` compute `res, k` using hints such that `res + k * modulus = el`
2) assert in-circuit that `api.AssertIsEqual(x, api.Add(res, api.Mul(k. modulus)))`

I leave the implementation do you for practice ;)

**Comment by gbotrel** [comment]
re: using circom circuit with gnark;  as @ivokub said, possible, but not trivial to implement I think. The mapping of the constraints should be straightforward, but the "solver hint" part I have no idea.

**Reply by BeratOz01** [reply]
Okay, how about modulus operation with two frontend.Variables ?

**Reply by BeratOz01** [reply]
Any updates on modulus ?
